jupyter_book/build.py

```diff
@@ -51,10 +51,11 @@ def _prepare_toc(toc):
             new_toc.append(section)
             new_toc.extend(subsections)
 
-    # Omit items that don't have URLs (like dividers) or have an external link
+    # Omit items that don't have URLs (like dividers), have an external link,
+    # or are internal links (have an anchor)
     return [
         item for item in new_toc
-        if 'url' in item and not item.get('external', False)
+        if 'url' in item and not item.get('external', False) and item.get('anchor') is None
```

Comment:
here rather than removing anything that has an anchor, why don't we instead just remove any redundant items from this returned list. So something like

```python
all_pages = [
        item for item in new_toc
        if 'url' in item and not item.get('external', False)
]
used_urls = []
pages_to_build = []
for page in all_pages:
    if page.url in used_urls:
        pass
    pages_to_build.append(page)
    used_urls.append(page.url)
return pages_to_build
```

I think that'd ensure that we don't double-build pages, and would also include pages if they *only* showed up w/ an anchor link

Reply:
I agree, the mechanism you describe removes the limitation that anchor links can only be subordinate to already-built pages. I think I was trying to be too cute: I had it as a goal not to clutter up the code too much because it was so easy to read!

It's too bad the `items` are dicts because if they were hashable I think the same behavior could be achieved by replacing the `all_pages` list comp with a set comp.

Reply:
I've incorporated your suggested change from this discussion. 

There are still a few outstanding issues:

 1. The anchor links don't work when they have apostrophe's in them (e.g. try clicking on the "Math Rendering" sidebar item on [my build of the jupyter-book website](https://rossbar.github.io/jupyterbook-site-test/intro.html)). I thought this might be a simple html-encoding issue, but unfortunately it seems like the `url_encode` liquid filter and anchorjs do not play nice with each other. I will look for a different solution.

 2. There is an annoying behavior in the sidebar: if you click a top-level link in the sidebar that has anchors, all of the anchor links light up blue (e.g. click on the "Frequently Asked Questions (FAQ)" link in the sidebar on on [my build of the jupyter-book website](https://rossbar.github.io/jupyterbook-site-test/intro.html)). I think this has to do with the fact that the `data-url` of the sidebar items in `sidebar.html` refers to the same top-level link even when an anchor is applied. Nothing is necessarily "broken", but it is ugly so I'd like to try to fix it.

One conclusion that I'm coming to by working on this is that I find liquid to be very cumbersome. I have no problem working on this and will continue to do so, but if you think that my approach is opening up too much of a can-of-worms please let me know!

Reply:
> One conclusion that I'm coming to by working on this is that I find liquid to be very cumbersome.

One of the reasons it's worth exploring Sphinx + Jinja? At least then we could control some of the logic within Python instead of just within the template language. For example, [here the pydata sphinx theme defines its own python function to get the nav object](https://github.com/pandas-dev/pydata-bootstrap-sphinx-theme/blob/master/pandas_sphinx_theme/__init__.py#L64) and then [it loads the nav object into Jinja here](https://github.com/pandas-dev/pydata-bootstrap-sphinx-theme/blob/master/pandas_sphinx_theme/docs-navbar.html#L16)